segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/AsyncStorageWrapper.java

```diff
@@ -108,62 +128,127 @@ public boolean supportsTruncation() {
+            } else {
+                // We need to wait on these futures to complete before executing ours.
+                result = CompletableFuture.allOf(futures)
+                        .handleAsync((r, ex) -> execute(operation, taskId, segmentNames), this.executor);
             }
-        });
+
+            // Update the last task for each involved segment to be this so that future tasks can be properly sequenced.
+            RunningTask t = new RunningTask(taskId, result);
+            for (String s : segmentNames) {
+                this.lastTasks.put(s, t);
+            }
+        }
 
         return result;
     }
 
     /**
      * Executes the given RunnableWithException asynchronously and returns a CompletableFuture that will be completed
      * when the Runnable completes.
+     * @param operation    The RunnableWithException to execute.
+     * @param segmentNames The names of the Segments involved in this operation (for sequencing purposes).
      */
-    private CompletableFuture<Void> runAsync(RunnableWithException operation) {
-        Exceptions.checkNotClosed(this.closed.get(), this);
-        CompletableFuture<Void> result = new CompletableFuture<>();
-        this.executor.execute(() -> {
-            try {
-                operation.run();
-                result.complete(null);
-            } catch (Throwable e) {
-                result.completeExceptionally(e);
+    private CompletableFuture<Void> runAsync(RunnableWithException operation, String... segmentNames) {
+        return supplyAsync(() -> {
+            operation.run();
+            return null;
+        }, segmentNames);
+    }
+
+    /**
+     * Executes the given Callable synchronously and invokes cleanup when done.
+     *
+     * @param operation    The Callable to execute.
+     * @param taskId       The id of the current task to be used for cleanup purposes.
+     * @param segmentNames The names of the Segments involved in this operation (for sequencing purposes).
+     */
+    @SneakyThrows(Exception.class)
+    private <R> R execute(Callable<R> operation, int taskId, String[] segmentNames) {
+        try {
+            return operation.call();
+        } finally {
+            cleanupIfNeeded(taskId, segmentNames);
+        }
+    }
+
+    private void cleanupIfNeeded(int taskId, String[] segmentNames) {
+        synchronized (this.lastTasks) {
+            ArrayList<String> toRemove = new ArrayList<>();
+            for (String s : segmentNames) {
+                // A segment entry can be safely cleaned up if the last registered task has the same id as the one
+                // we got in this method (that means no more tasks have been added).
+                val task = this.lastTasks.get(s);
+                if (task != null && task.taskId == taskId) {
+                    toRemove.add(s);
+                }
             }
-        });
 
-        return result;
+            toRemove.forEach(this.lastTasks::remove);
```

Comment:
Just do this is the above for loop. It' under the same lock and toRemove is not used for any other reason.

Reply:
Fixed. This was a leftover from when I was looping over the map itself.